nft_ingester/src/backfiller.rs

```diff
@@ -0,0 +1,702 @@
+            })
+            .unwrap();
+
+        // Get RPC URL.
+        let rpc_url = config
+            .rpc_config
+            .get(&*RPC_URL_KEY)
+            .and_then(|u| u.clone().into_string())
+            .ok_or(IngesterError::ConfigurationError {
+                msg: format!("RPC URL missing: {}", RPC_URL_KEY),
+            })
+            .unwrap();
+
+        // Instantiate RPC client.
+        let rpc_client = RpcClient::new_with_commitment(rpc_url, CommitmentConfig::processed());
+
+        // Instantiate messenger.
+        let mut messenger = T::new(config.messenger_config).await.unwrap();
+        messenger.add_stream(TRANSACTION_STREAM).await;
+        messenger.set_buffer_size(TRANSACTION_STREAM, 5000).await;
+
+        Self {
+            db,
+            listener,
+            rpc_client,
+            messenger,
+            failure_delay: INITIAL_FAILURE_DELAY,
+        }
+    }
+
+    /// Run the backfiller task.
+    async fn run(&mut self) {
+        loop {
+            match self.get_trees_to_backfill().await {
+                Ok(backfill_trees) => {
+                    if backfill_trees.len() == 0 {
+                        // If there are no trees to backfill, wait for a notification on the channel.
+                        let _notification = self.listener.recv().await.unwrap();
+                    } else {
+                        println!("New trees to backfill: {}", backfill_trees.len());
+
+                        // First just check if we can talk to an RPC provider.
+                        match self.rpc_client.get_version().await {
+                            Ok(version) => println!("RPC client version {version}"),
+                            Err(err) => {
+                                println!("RPC client error {err}");
+                                self.sleep_and_increase_delay().await;
+                                continue;
+                            }
+                        }
+
+                        for backfill_tree in backfill_trees {
+                            // Get the tree out of nested structs.
+                            let tree = backfill_tree.unique_tree.tree;
+                            let tree_str = bs58::encode(&tree).into_string();
+                            println!("Backfilling tree: {tree_str}");
+
+                            // Call different methods based on whether tree needs to be backfilled
+                            // completely from seq number 1 or just have any gaps in seq number
+                            // filled.
+                            let result = if backfill_tree.backfill_from_seq_1 {
+                                self.backfill_tree_from_seq_1(&tree).await
+                            } else {
+                                self.fetch_and_plug_gaps(&tree).await
+                            };
+
+                            match result {
+                                Ok(opt_max_seq) => {
```

Comment:
Yes correct we don't assume we fixed everything, and yes we saved the max number that we did fix up to.

Stuff gets added above this `seq` by normal geyser moving forward, and stuff gets added below this `seq` because of the results of backfilling itself (see other comments for more explanation on that phenomenon).